src/grpc/projects.rs

```diff
@@ -105,328 +109,188 @@ impl ProjectService for ProjectServiceImpl {
-        //         return Err(tonic::Status::permission_denied("Not allowed."));
-        //     }
-        // };
-        // let client = self.database.get_client().await.map_err(|e| {
-        //     log::error!("{}", e);
-        //     tonic::Status::unavailable("Database not avaliable.")
-        // })?;
-
-        // let dataclass = inner_request.data_class.try_into().map_err(|e| {
-        //     log::error!("{}", e);
-        //     tonic::Status::internal("DataClass conversion error.")
-        // })?;
-        // let old_class: i32 = Object::get(object_id, &client)
-        //     .await
-        //     .map_err(|e| {
-        //         log::error!("{}", e);
-        //         tonic::Status::internal("Database transaction failed.")
-        //     })?
-        //     .ok_or(tonic::Status::internal("Database transaction failed."))?
-        //     .data_class
-        //     .into();
-        // if old_class > inner_request.data_class {
-        //     return Err(tonic::Status::internal("Dataclass can only be relaxed."));
-        // }
-        // Object::update_dataclass(object_id, dataclass, &client)
-        //     .await
-        //     .map_err(|e| {
-        //         log::error!("{}", e);
-        //         tonic::Status::aborted("Database update failed.")
-        //     })?;
-        // let object = Object::get_object_with_relations(&object_id, &client)
-        //     .await
-        //     .map_err(|e| {
-        //         log::error!("{}", e);
-        //         tonic::Status::aborted("Database update failed.")
-        //     })?;
-        // let project = Some(object.try_into().map_err(|e| {
-        //     log::error!("{}", e);
-        //     tonic::Status::aborted("Database request failed.")
-        // })?);
-        // Ok(tonic::Response::new(UpdateProjectDataClassResponse {
-        //     project,
-        // }))
-    }
-    async fn get_projects(
-        &self,
-        _request: Request<GetProjectsRequest>,
-    ) -> Result<Response<GetProjectsResponse>> {
-        todo!()
+        log_received!(&request);
+
+        let token = tonic_auth!(
+            get_token_from_md(request.metadata()),
+            "Token authentication error."
+        );
+
+        let request = DataClassUpdate::Project(request.into_inner());
+        let collection_id = tonic_invalid!(request.get_id(), "Invalid project id");
+        let ctx = Context::res_ctx(collection_id, DbPermissionLevel::WRITE, true);
+
+        tonic_auth!(
+            self.authorizer.check_permissions(&token, vec![ctx]).await,
+            "Unauthorized"
+        );
+
+        let project = tonic_internal!(
+            self.database_handler.update_dataclass(request).await,
+            "Internal database error."
+        );
+        self.cache
+            .update_object(&project.object.id, project.clone());
+        let project: generic_resource::Resource =
+            tonic_internal!(project.try_into(), "Collection conversion error");
```

Comment:
Copy & paste only partially adapted.

```suggestion
        let project_id = tonic_invalid!(request.get_id(), "Invalid project id");
        let ctx = Context::res_ctx(project_id, DbPermissionLevel::WRITE, true);

        tonic_auth!(
            self.authorizer.check_permissions(&token, vec![ctx]).await,
            "Unauthorized"
        );

        let project = tonic_internal!(
            self.database_handler.update_dataclass(request).await,
            "Internal database error."
        );
        self.cache
            .update_object(&project.object.id, project.clone());
        let project: generic_resource::Resource =
            tonic_internal!(project.try_into(), "Project conversion error");
```